docs/spec/tuples.rst

```diff
@@ -0,0 +1,155 @@
+Tuples
+======
+
+The ``tuple`` class has some special behaviors and properties that make it
+different from other classes from a typing perspective. The most obvious
+difference is that ``tuple`` is variadic -- it supports an arbitrary number
+of type arguments. At runtime, the sequence of objects contained within the
+tuple is fixed at the time of construction. Elements cannot be added, removed,
+reordered, or replaced after construction. These properties affect subtyping
+rules and other behaviors as described below.
+
+
+Tuple Type Form
+---------------
+
+The type of a tuple can be expressed by listing the element types. For
+example, ``tuple[int, int, str]`` is a tuple containing an ``int``, another
+``int``, and a ``str``.
+
+The empty tuple can be annotated as ``tuple[()]``.
+
+Arbitrary-length homogeneous tuples can be expressed using one type and
+ellipsis, for example ``tuple[int, ...]``. This type is equivalent to a union
```

Comment:
I think technically you have to say "and **an** ellipsis".

Reply:
Good catch. I've pushed a fix.

This text was copied directly from the [original language in PEP 484](https://peps.python.org/pep-0484/#the-typing-module). DIdn't you write much of that PEP? :)

Reply:
> I also wonder if we ought to say something about sequence unpacking?

I agree the spec should eventually say something about tuple unpacking. Note that tuple unpacking is different from general sequence unpacking, especially in cases where a tuple has a known length.

I started to write a section about tuple unpacking behaviors, but I decided to defer that discussion for now. I think the topic will be somewhat controversial. It's also intertwined with an even more controversial topic: how to treat an unpacked tuple in a function call (e.g. `func(*my_tuple)`) when the length of the tuple isn't known or the function includes an `*args` parameter. The behaviors here differ significantly across type checkers.

This is on a long list of issues I'm tracking. We'll get to it eventually.